File: 32. Longest Valid Parentheses/32. Longest Valid Parentheses.rb
# @param {String} s
# @return {Integer}
def longest_valid_parentheses(s)
    max_len = 0
    stack = [-1]

    s.each_char.with_index do |char, i|
        if char == '('
            stack.push(i)
        else
            stack.pop
            if stack.empty?
                stack.push(i)
            else
                max_len = [max_len, i - stack.last].max
            end
        end
    end

    max_len
end

File: 32. Longest Valid Parentheses/32. Longest Valid Parentheses.go
func longestValidParentheses(s string) int {
    maxLen := 0
    stack := []int{-1}

    for i := 0; i < len(s); i++ {
        if s[i] == '(' {
            stack = append(stack, i)
        } else {
            stack = stack[:len(stack)-1]
            if len(stack) == 0 {
                stack = append(stack, i)
            } else {
                if i - stack[len(stack)-1] > maxLen {
                    maxLen = i - stack[len(stack)-1]
                }
            }
        }
    }

    return maxLen
}